module_utils/network_lsr/argument_validator.py

```diff
@@ -633,6 +704,177 @@ def _validate_post(self, value, name, result):
+            result["from"] = network_from["address"]
+            result["family"] = family
+        if result["to"]:
+            network_to = result["to"]
+            family = network_to["family"]
+            result["to"] = network_to["address"]
+            result["family"] = family
+
+        if result["action"] == "to-table":
+            if result["table"] is None:
+                raise ValidationError(
+                    name,
+                    "missing table for the routing rule",
+                )
+
+        if result["from_len"] == 0:
+            if result["from"] is not None:
+                raise ValidationError(
+                    name,
+                    "the prefix length for from/src cannot be zero when from/src is specified",
+                )
+        elif result["from"] is None:
+            raise ValidationError(name, "missing from/src for a non zero prefix-length")
+        elif result["from_len"] > (32, 128)[result["family"] == socket.AF_INET6]:
+            raise ValidationError(name, "the prefix length for from/src is invalid")
+
+        if result["to_len"] == 0:
+            if result["to"] is not None:
+                raise ValidationError(
+                    name,
+                    "the prefix length for to/dst cannot be zero when to/dst is specified",
+                )
+        elif result["to"] is None:
+            raise ValidationError(name, "missing to/dst for a non zero prefix-length")
+        elif result["to_len"] > (32, 128)[result["family"] == socket.AF_INET6]:
+            raise ValidationError(name, "the prefix length for to/dst is invalid")
+
+        if result["fwmark"] is not None:
+            if result["fwmask"] is None:
+                raise ValidationError(
+                    name,
+                    "the fwmark requires fwmask to be set",
+                )
+
+        if result["fwmask"] is not None:
+            if result["fwmark"] is None:
+                raise ValidationError(
+                    name,
+                    "the fwmask requires fwmark to be set",
+                )
+        if result["iif"] is not None:
+            if not Util.ifname_valid(result["iif"]):
+                raise ValidationError(
+                    name,
+                    "the incoming interface '{0}' specified in the routing rule is invalid interface_name.".format(
+                        result["iif"]
+                    ),
+                )
+
+        if result["oif"] is not None:
+            if not Util.ifname_valid(result["oif"]):
+                raise ValidationError(
+                    name,
+                    "the outgoing interface '{0}' specified in the routing rule is invalid interface_name.".format(
+                        result["oif"]
+                    ),
+                )
+        if result["sport"]:
+            if (
+                result["sport"]["range_start"] == 0
+                and result["sport"]["range_end"] != 0
```

Comment:
what is the meaning of this check? `ArgValidatorRange` should already sufficiently validate the range.

Reply:
NM validated this way, https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/blob/main/src/libnm-core-impl/nm-setting-ip-config.c#L2622

Reply:
I mean that the constancy rules (validate) should be handled by where the error happens. That is, `ArgValidatorRange` needs to validate that the range is valid, and not `ArgValidatorIPRoutingRules`.

The result of `ArgValidatorRange` should be either "None" or a valid range (that is,one that satisfies the min/max values and the min<max rule).

`_rr_xport_range_valid()` does this check, because `NMIPRoutingRule` also needs to express to specify no range (`None`) or a valid range. Since `0` is not a valid port anyway (outside of the valid range), it uses that value to encode a missing range. Of course, setting `dport_start=0` && `dport_end!=0` would be meaningless/inconsistent, so that is rejected as invalid.

`ArgValidatorRange` encodes a missing range entirely differently (`None`).

Reply:
See also `ArgValidatorRange("sport", val_min=0, val_max=65534)`.

The `val_min` of a port must be 1, not 0.
A missing port is encoded by not specifying the range (and `ArgValidatorRange` will represent that as `None`).